First cut

eskatos opened the pull request:
The content depends on Gradle 4.8 RC1.
`:checkLinks` fails as the documentation links aren't valid for the release-candidate.
WIP
**Do not merge**!

eriwen commented on contents/index.adoc:
```diff
@@ -1,22 +1,936 @@
-== CONTENT_TBC
+[TIP]
+The Gradle link:{user-manual}[user manual] and link:{guides}[guides] contain build script excerpts that demonstrate both the Groovy DSL and the Kotlin DSL.
+This is the best place where to find how to do this and what with each DSL  ; and it covers all Gradle features from link:{user-manual}plugins.html[using plugins] to link:{user-manual}customizing_dependency_resolution_behavior.html[customizing the dependency resolution behavior].
```
❓ are the extra spaces before the `;` intentional?

eskatos replied:
Unintentional, fixing

eriwen commented on contents/index.adoc:
```diff
@@ -1,22 +1,936 @@
+:figure-caption!:
 
-*INTRO_TBC*
+In this guide we will look at the main differences between Gradle's Groovy DSL and Kotlin DSL by comparing build scripts doing common things, introducing the main idioms of the Kotlin DSL.
```
❓ would it better to summarize this guide by making these 3 things bullet points?

eskatos replied:
I'm balanced, your call

eriwen commented on contents/index.adoc:
```diff
@@ -1,22 +1,936 @@
+--
+*Applying Gradle best practices will make it much easier with the Kotlin DSL*
+
+Please refer to the link:{user-manual}authoring_maintainable_build_scripts.html[Authoring Maintainable Build Scripts] and link:{user-manual}organizing_build_logic.html[Organizing Build Logic] chapters.
```
"chapters" => "user manual chapters"?

eriwen commented on contents/index.adoc:
```diff
@@ -1,22 +1,936 @@
+*Applying Gradle best practices will make it much easier with the Kotlin DSL*
+
+Please refer to the link:{user-manual}authoring_maintainable_build_scripts.html[Authoring Maintainable Build Scripts] and link:{user-manual}organizing_build_logic.html[Organizing Build Logic] chapters.
+In general, follow the advices given in the Gradle link:{user-manual}[user manual] and link:{guides}[guides].
```
Native English writers prefer "advice" over "advices"

eriwen commented on contents/index.adoc:
```diff
@@ -1,22 +1,936 @@
-////
 
-*TBC*
+== Groovy meet Kotlin
```
❗️  This title does not make it clear what this section is about. We might consider something like "Groovy vs. Kotlin DSL tradeoffs" or "Which DSL should you choose? Groovy or Kotlin"

eriwen commented on contents/index.adoc:
```diff
@@ -1,22 +1,936 @@
+But, the declarative nature of the `plugins {}` block enables the Kotlin DSL to provide type safe discoverable accessors to what the applied plugins contribute (extensions, configurations etc...).
+
+
+=== Applying core plugins
```
❓ Should we consider combining "Applying core plugins" and "Applying external plugins" into one example?

eriwen replied:
We seem to do this in the "Configuring plugins" section.

eskatos replied:
Yep, we could combine these two sections. But I think that a single sample would be crowded, especially with both a `buildscript {}` and a `plugins {}` block. I kinda like the fact that the current text goes step by step. No strong opinion though.

eriwen commented on contents/index.adoc:
```diff
@@ -1,22 +1,936 @@
+There may be more than one object for some plugins.
+
+
+== Prefer the `plugins {}` block
```
❗️ I may be missing it, but I don't think we touch on why one should prefer the `plugins {}` block here. I'm sure there's a bit in the best practices doc about this we can point to, but I cannot find it right now.

Alternatively, we can change this title to "Keeping your Kotlin Gradle script declarative"

eskatos replied:
I don't like this title either. I like your alternative proposal, it emphasize the importance of being declarative. I'll change it to `Keeping Kotlin DSL scripts declarative` for now.

eriwen commented on contents/index.adoc:
```diff
@@ -1,22 +1,936 @@
+    // ...
+    resolutionStrategy {
+        eachPlugin {
+            if (requested.id.id == "com.android.application") {
```
❓ `id.id` is that right?

eriwen commented on contents/index.adoc:
```diff
@@ -1,22 +1,936 @@
+tasks.getByName<Jar>("jar").archiveName = "foo.jar"
+----
+
+Note that specifying the type of the task explicitly is necessary. Otherwise, the script won't compile because the inferred type of `jar` will be `Task`, and the `archiveName` property is specific to the `Jar` task type.
```
❓ Should we use a `NOTE:` admonition here?

eriwen commented on contents/index.adoc:
```diff
@@ -1,22 +1,936 @@
+
+[source.multi-language-sample, groovy]
+----
+test.doLast {
```
I think we should normalize to users _always_ declaring the task type, thus reducing the number of DSL constructs users will encounter.

eriwen commented on contents/index.adoc:
```diff
@@ -1,22 +1,936 @@
+}
+----
+
+But there is another idiomatic way to configure tasks: using Kotlin link:https://kotlinlang.org/docs/reference/delegated-properties.html[delegated properties].
```
❗️ I suggest that we establish this as the 1 way to configure tasks because:

 * It holds true to the "higher ceremony for greater type safety" compromise explained earlier
 * It improves auto-completion
 * Other build authors may come along needing to reference the task at another time. 
 * I will suggest we make this consistent in "Creating tasks" by having exactly 1 established way using `tasks.creating`.
 * Up front establishing one way to do this reduces mental overhead for build script maintainers.

I think this is so important that I would even suggest _deprecating the other ways of declaring and creating tasks_. 

What do you think @eskatos and @bamboo? Please hit me up on Slack if that's a better place to discuss.

eriwen commented on contents/index.adoc:
```diff
@@ -1,22 +1,936 @@
+}
+----
+
+Note that if you do not use the `plugins {}` block to apply your plugins, then accessors to configurations contributed by the applied plugins won't be available. You'll then have to address them by name:
```
❓ What does "then accessors to configurations contributed by the applied plugins won't be available" mean in less technical language?

eskatos replied:
I rephrased to `Note that if you do not use the `plugins {}` block to apply your plugins, then Kotlin extensions to reference the configurations contributed by the applied plugins won't be available.`

eriwen commented on contents/index.adoc:
```diff
@@ -1,22 +1,936 @@
+}
+----
+
+or bring them into scope thanks to Kotlin delegated properties:
```
❗️ We mention "delegated properties" so often in this guide that we might even include a little section on what they are.

eskatos replied:
The Kotlin reference documentation on delegated properties is linked the first time they are mentioned. I don't think we should paraphrase the Kotlin ref doc.
Maybe we should link to it each time they are mentioned, in case people simply go over the first one, wdty?

eriwen commented on contents/index.adoc:
```diff
@@ -1,22 +1,936 @@
+Both share the same ultimate goal though: provide means to interact with the dynamic and extensible Gradle model, and runtime.
+
+It takes more ceremony to do dynamic things with the Kotlin DSL.
+On the other hand it provides more safety and more tooling.
```
❓ Should we mention interactive debugging here? I recognize both are supported as of IDEA 2018.2, but is Kotlin better and in what ways?

eskatos replied:
Yep, debugging is now supported for both Groovy and Kotlin scripts.
I didn't assess the difference in debugging support, hence not mentioning it.
That's now a detail we can add as an enhancement later.

eriwen commented on contents/index.adoc:
```diff
@@ -1,22 +1,936 @@
+
+// TODO redact
+
+Both scripts using the Kotlin DSL, `\*.gradle.kts`, and scripts using the Groovy DSL, `*.gradle`, can participate in the same build. Gradle plugins implemented in `./buildSrc`, an included build or externally fetched can use any JVM language. This makes it possible to migrate progressively, piece by piece, without blocking your team.
```
I have suggested to users that one strategy is to break up build.gradle into separate Gradle script(s) and apply them inside a root `build.gradle.kts`.

eriwen commented on contents/index.adoc:
```diff
@@ -1,22 +1,936 @@
+--
+Depending on your build structure you might be interested by the following samples built with the Kotlin DSL:
+
+* link to modularity sample
```
I think we should link to samples early and often here.

eriwen commented on contents/index.adoc:
```diff
@@ -1,22 +1,936 @@
+----
+
+
+== Configuring plugins
```
❗️ We might consider linking to more concrete samples in this and later subsections.

eskatos replied:
We can add more links to concrete examples over time from guides and the user manual once they cover the Kotlin DSL.

eriwen commented on contents/index.adoc:
```diff
@@ -1,22 +1,936 @@
+If you want to read on Kotlin the language before getting started or if you need some reference documentation along the way, the https://kotlinlang.org/docs/reference/[reference documentation] is the place to go.
+The https://kotlinlang.org/docs/tutorials/koans.html[Kotlin Koans] provide a fun way to learn Kotlin.
+
+The Kotlin language being more strict than Groovy also means that the Gradle Kotlin DSL is more strict than the Gradle Groovy DSL.
```
❗️ Not sure where to put this comment, but we might start with suggesting users upgrade to the latest Gradle version.

eskatos replied:
Good point! I added it as a good first step wrt applying best practice near the top of the guide.

eskatos commented:
Hey Eric,
Thank you for the review!
I'd like give the `Multi project builds`, `Migration strategies` and `Tips and troubleshooting` sections some more love before merging this PR. Feel free to open PR's against this one.
Oh and, we'll need 4.8 GA to make CI pass on this anyway.
I'll hit you on slack to discuss domain object containers usages (`tasks` being one of them).

pioterj commented:
@eskatos Some generic feedback from my side after skimming through the guide:

- "Groovy vs. Kotlin DSL tradeoffs" could emphasize practical advantages of Kotlin DSL more, especially vastly improved IDE support. It only mentions "more tooling" in this context. 
- I'd consider skipping coverage of imperative apply syntax. We don't want people to use it, especially with Kotlin DSL. Why not just say that for the best experience you should use the plugins block? And just show how to accomplish that with Android plugin rather than elaborating on `configure<AppExtension>`?
- We show multiple ways to create a task without much context about pros and cons on each way. Shoudn't we agree on a single preferred way to create a task?

eskatos commented:
@pioterj, thank you for the review!

Good point about IDE support, I added some words to that effect.

I think skipping imperative use cases would make this *migration* guide less actionable hence less useful. I also think there's enough emphasis on declarative scripts and Gradle best practices.

I simplified both sections on tasks, still subject to change.

pioterj commented:
> I think skipping imperative use cases would make this migration guide less actionable hence less useful.

@eskatos Could you explain why? To me the use case is applying a plugin. Eventually we want to have a single way to do this. I was actually wondering whether it would make sense to deprecate the old way in the Kotlin DSL already.

eskatos commented:
@pioterj, I agree with your comments in a broader scope but a migration might not be a one step thing and I think knowing both ways will be handy for migrating existing groovy builds step by step.

eriwen commented on contents/index.adoc:
```diff
@@ -1,22 +1,768 @@
+Both link:{kotlin-reference}java-interop.html[calling Java from Kotlin] and link:{kotlin-reference}java-to-kotlin-interop.html[calling Kotlin from Java] are very well covered in the Kotlin reference documentation.
+
+The same mostly apply to interoperability with Groovy code.
+There are some gotchas though.
```
❓ I wonder if this sentence provides much value.

eriwen commented on contents/index.adoc:
```diff
@@ -1,22 +1,768 @@
+[NOTE]
+Gradle plugins written in any language should prefer the typed `Action<T>` that is automatically mapped to Groovy closures and Kotlin lambdas.
+
+Some plugins expect simple closures, here with the link:{plugin-portal}plugin/com.jfrog.bintray[Bintray] plugin:
```
❓ should "here with the" be rephrased as "as with the"?

eriwen commented on contents/index.adoc:
```diff
@@ -1,22 +1,768 @@
+You also took a glance at possible migration strategies in the light of the structure of builds.
+Last but not least you learnt how the two DSLs inter-operate.
+
+Next steps:
```
Perhaps a subheader here? `===`

eriwen commented on contents/index.adoc:
```diff
@@ -1,22 +1,768 @@
+
+Next steps:
+
+* The Gradle link:{user-manual}[user manual] and link:{guides}[guides] contain build script excerpts that demonstrate both the Groovy DSL and the Kotlin DSL.
```
This will not be true if we publish this guide first, so we might consider omitting this for now.

eriwen commented on samples/code/configurations-and-dependencies/imperative-string-reference.gradle.kts:
```diff
@@ -0,0 +1,9 @@
+apply(plugin = "java")
+dependencies {
+    "implementation"("com.example:lib:1.1")
```
Aside: Whoa, so you have to use one of these strategies if you use the `apply()` syntax? I know I read this, but it seems rough.

eskatos commented on contents/index.adoc:
```diff
@@ -1,22 +1,768 @@
+
+On top of that, the following conventions apply for better IDE support:
+
+* scripts applied to `Settings` should be named `*.settings.gradle.kts`
```
> ❓ Should be or must be?

Should be, for better IDE support as mentioned above.

> What does "scripts applied to Settings" mean?

**settings.gradle.kts**
```kotlin
apply(from = "other.settings.gradle.kts")
```

**other.settings.gradle.kts**
```kotlin
// this script is applied to Settings
```
